riptable/rt_str.py

```diff
@@ -713,8 +715,97 @@ def regex_match(self, regex):
         regex = re.compile(regex)
         vmatch = np.vectorize(lambda x: bool(regex.search(x)))
         bools = vmatch(self.backtostring)
+
         return bools
 
+    def extract(self, regex: str, expand: Optional[bool] = None,
+                fillna: str = '', names=None, apply_unique: bool = True):
+        '''
+        Extract one or more pattern groups into a Dataset or FastArray.
+        For one capture group the default is to return a FastArray
+        but this can be overridden by passing expand=True or by providing names.
+        Column names can be specified within the regex using (?P<name>) in the search group(s)
+        or by passing the names argument which may be more convenient.
+
+        Parameters
+        ----------
+        regex - Contains the patterns to search for
+        expand - set to True to return a Dataset for a single capture group.
+        fillna - Used for rows where no regex does not match
+        names - Optional list of strings provides keys for resultant dataset
+        apply_unique - When True we apply the regex to the unique values and then expand using the reverse index.
+        This is optimal for repetitive data and benign for unique or close highly non-repetitive data
+
+        Examples
+        --------
+        >>> osi = rt.FastArray(['SPX UO 12/15/23 C5700', 'SPXW UO 09/17/21 C3650'])
+        >>> osi.str.extract('\w+')
+        FastArray([b'SPX', b'SPXW'], dtype='|S4')
+        >>> osi.str.exract('(?P<root>\w+)')
+        #   root
+        -   ----
+        0   SPX
+        1   SPXW
+        >>> osi.str.exract('(\w+).* (\d{2}/\d{2}/\d{2})', names=['root', 'expiration'])
+        #   root   expiration
+        -   ----   ----------
+        0   SPX    12/15/23
+        1   SPXW   09/17/21
+        >>> osi.str.extract('\w+W', expand=True)
+        #   group_0
+        -   -------
+        0
+        1   SPXW
+        '''
+        kwargs = dict(expand=expand, fillna=fillna, names=names, apply_unique=apply_unique)
+        if apply_unique:
+            kwargs['apply_unique'] = False
+            unique_values, index = unique(self.backtostring, return_inverse=True)
+            result = unique_values.str.extract(regex, **kwargs)
+            return result[index] if isinstance(result, FastArray) else result[index, :]
+
+        if not isinstance(regex, bytes):
+            regex = bytes(regex, 'utf-8')
+        compiled = re.compile(regex)
+
+        ngroups = compiled.groups
+        if ngroups == 0:
+            # convenience where we treat the entire pattern as a capture group
+            return self.extract(f'({regex.decode()})', **kwargs)
+
+        # expand defaults to False if we have one capture group and do not specify names
+        if expand is None:
+            expand = ngroups > 1 or names is not None
+        elif expand is False and ngroups > 1:

```

Comment:
Please simplify ``expand is False`` to ``not expand`` here. (Or, if you want to be strict about the comparison instead of accepting falsey values, please use ``expand == False`` -- ``expand is False`` relies on CPython implementation details.)

Reply:
Agreed. This was intentional as I wanted want to distinguish between `expand=None` but I later did that more explicitly above.